File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        columns = len(grid[0])
        fresh = 0
        queue = []
        
        for i in range(rows):
            for j in range(columns):
                cell = grid[i][j]
                if (cell == 0):
                    continue
                elif (cell == 1):
                    fresh += 1
                    # oranges += 1
                else:
                    queue.append((i, j))
                    # oranges += 1


        def isValid(i, j):
            if (i < 0 or j < 0 or i == rows or j == columns or grid[i][j] != 1):
                return False
            return True

        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        minutes = 0
        while (queue and fresh > 0):
            size = len(queue)

            for _ in range(size):
                i, j = queue.pop(0)

                for x, y in directions:
                    new_x = x + i
                    new_y = y + j

                    if (isValid(new_x, new_y)):
                        queue.append((new_x, new_y))
                        grid[new_x][new_y] = 2
                        fresh -= 1
            minutes += 1
            if (fresh <= 0):
                break

        if (fresh > 0):
            return -1
        else:
            return minutes

File: 1036-rotting-oranges/rotting-oranges.js
var orangesRotting = function(grid) {
    let rows = grid.length
    let columns = grid[0].length
    let fresh = 0
    let queue = []

    function isValid(i, j) {
        if (i < 0 || j < 0 || i == rows || j == columns) {
            return false
        }
        return true
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            cell = grid[i][j]
            if (cell == 0) continue
            if (cell == 1) fresh += 1
            else queue.push([i, j])
        }
    }
    let directions = [[-1, 0], [0, 1], [1, 0], [0, -1]]

    let minutes = 0
    while (queue.length && fresh > 0) {
        let size = queue.length

        for (let k = 0; k < size; k++) {
            let [i, j] = queue.shift()
            
            for (let [x, y] of directions) {
                let row = i + x
                let col = j + y

                if (!isValid(row, col) || grid[row][col] != 1) continue
                grid[row][col] = 2
                fresh -= 1
                queue.push([row, col])
            }
        }
        minutes += 1
    }

    return fresh == 0 ? minutes : -1


};